### PR TITLE
[Compiler] Add Catalyst adjoint

### PR DESCRIPTION
**Pull request by rmoyard**
**Description of the Change:**

This PR adds support for QJIT compatible qml.adjoint.

### CODE REVIEW

**Comment by rmoyard** [comment]
[sc-46403]

**Review by maliasadi — CHANGES_REQUESTED** [review]
Thanks @rmoyard :tada:

File: pennylane/ops/op_math/adjoint.py [hunk 1]
```diff
@@ -117,6 +144,12 @@ def circuit(a):
         Adjoint(S)(wires=[0])
 
     """
+    if active_jit := compiler.active_compiler():
+        if lazy is False:
+            raise RuntimeError("Lazy kwarg is not support with qjit.")
```
Comment (dime10) [inline pennylane/ops/op_math/adjoint.py:149]: For all Catalyst related errors during program capture or compilation:
```suggestion
            raise CompileError("Lazy kwarg is not support with qjit.")
```

File: pennylane/ops/op_math/adjoint.py [hunk 2]
```diff
@@ -101,6 +102,32 @@ def circuit(a):
     >>> print(qml.draw(circuit)(0.2))
     0: ──RX(0.20)──SX──SX†──RX(0.20)†─┤  <Z>
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/ops/op_math/adjoint.py:104]: Updated let me know if you have more comments

File: pennylane/ops/op_math/adjoint.py [hunk 3]
```diff
@@ -44,7 +47,16 @@ def adjoint(fn, lazy=True):
...
+        The adjoint and inverse are identical for unitary gates, but not in general. For example, quantum channels and
+        observables may have different adjoint and inverse operators.
+
+    .. note::
+
+        When used with :func:`~.qjit`, this function only supports the Catalyst compiler.
+        Please see :func:`catalyst.adjoint` for more details.
+        Please see the Catalyst :doc:`quickstart guide <catalyst:dev/quick_start>`,
+        as well as the :doc:`sharp bits and debugging tips <catalyst:dev/sharp_bits>`
+        page for an overview of the differences between Catalyst and PennyLane.
```
Comment (maliasadi) [inline pennylane/ops/op_math/adjoint.py:59]:
```suggestion
        When used with :func:`~.qjit`, this function only supports the Catalyst compiler.
        See :func:`catalyst.adjoint` for more details.

        Please see the Catalyst :doc:`quickstart guide <catalyst:dev/quick_start>`,
        as well as the :doc:`sharp bits and debugging tips <catalyst:dev/sharp_bits>`
        page for an overview of the differences between Catalyst and PennyLane.
```

File: pennylane/ops/op_math/adjoint.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (maliasadi) [reply pennylane/ops/op_math/adjoint.py:116]: Thanks!

File: pennylane/ops/op_math/adjoint.py [hunk 4]
```diff
@@ -21,13 +21,16 @@
...
+from pennylane.compiler.compiler import CompileError
 
 from .symbolicop import SymbolicOp
 
 
 # pylint: disable=no-member
 def adjoint(fn, lazy=True):
-    """Create the adjoint of an Operator or a function that applies the adjoint of the provided function.
+    """A :func:`~.qjit` compatible adjoint transformation that creates the adjoint of an Operator or a function
+    that applies the adjoint of the provided function.
```
Comment (josh146) [inline pennylane/ops/op_math/adjoint.py:33]:
```suggestion
```

File: pennylane/ops/op_math/adjoint.py [hunk 5]
```diff
@@ -21,13 +21,16 @@
...
 
     Args:
         fn (function or :class:`~.operation.Operator`): A single operator or a quantum function that
```
Comment (josh146) [inline pennylane/ops/op_math/adjoint.py:36]: @rmoyard I can't comment on it, but for the `lazy` kwarg, I would add the follwing:

```rest
lazy=True (bool): If the transform is behaving lazily, all operations are
    wrapped in a ``Adjoint`` class and handled later. If ``lazy=False``,
    operation-specific adjoint decompositions are first attempted.
    This argument is not supported when used with :func:`~.qjit`.

**Review by josh146 — APPROVED** [review]
💯

File: pennylane/ops/op_math/adjoint.py (hunk 5 above)
Comment (rmoyard) [reply pennylane/ops/op_math/adjoint.py:36]: Thanks I will add that !

File: pennylane/ops/op_math/adjoint.py [hunk 6]
```diff
@@ -21,13 +21,15 @@
...
+from pennylane.compiler import compiler
+from pennylane.compiler.compiler import CompileError
 
 from .symbolicop import SymbolicOp
 
 
 # pylint: disable=no-member
 def adjoint(fn, lazy=True):
-    """Create the adjoint of an Operator or a function that applies the adjoint of the provided function.
+    """A :func:`~.qjit` compatible adjoint transformation that creates the adjoint of an Operator or a function
```
Comment (timmysilv) [inline pennylane/ops/op_math/adjoint.py:32]: I know it was repeated before but it stutters quite a bit imo
```suggestion
    """A :func:`~.qjit` compatible transformation that creates the adjoint of an Operator or a function
```

File: pennylane/ops/op_math/adjoint.py [hunk 7]
```diff
@@ -36,6 +38,7 @@ def adjoint(fn, lazy=True):
     Keyword Args:
         lazy=True (bool): If the transform is behaving lazily, all operations are wrapped in a ``Adjoint`` class
             and handled later. If ``lazy=False``, operation-specific adjoint decompositions are first attempted.
+            This argument is not supported when used with :func:`~.qjit`.
```
Comment (timmysilv) [inline pennylane/ops/op_math/adjoint.py:41]:
```suggestion
            Setting ``lazy=False`` is not supported when used with :func:`~.qjit`.
```

File: pennylane/ops/op_math/adjoint.py [hunk 8]
```diff
@@ -117,6 +158,12 @@ def circuit(a):
         Adjoint(S)(wires=[0])
 
     """
+    if active_jit := compiler.active_compiler():
+        if lazy is False:
+            raise CompileError("Lazy kwarg is not support with qjit.")
```
Comment (timmysilv) [inline pennylane/ops/op_math/adjoint.py:163]:
```suggestion
            raise CompileError("Setting lazy=False is not supported with qjit.")
```

**Review by timmysilv — APPROVED** [review]
a few doc comments but looks good to me!

**Review by maliasadi — APPROVED** [review]

**Pull request merged**